### PR TITLE
Fix outputs of Sequences and Maps exposure.

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**:
Make sure that all OrtValues are properly disposed of. This includes map and sequence ortvalues as well as sequence elements that are currently leaked as the DisposableList is currently assigned to NamedOnnxValue member which does not dispose anything.

**Motivation and Context**

### CODE REVIEW

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 1]
```diff
@@ -502,8 +502,7 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> RunWithBindingAnd
                     for (int i = 0; i < outputNames.Length; ++i)
                     {
                         var ortValue = ortValues.ElementAt(i);
-                        result.Add(DisposableNamedOnnxValue.CreateTensorFromOnnxValue(outputNames[i], ortValue.Handle));
-                        ortValue.Disown();
+                        result.Add(DisposableNamedOnnxValue.CreateTensorFromOnnxValue(outputNames[i], ortValue));
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:505]: >CreateTensorFromOnnxValue [](start = 60, length = 25)

We should call CreateFromOrtValue to allow for maps and sequences.

**Review by snnn — APPROVED** [review]

**Comment by hariharans29** [comment]
Are existing tests already covering this ?

**Comment by yuslepukhin** [comment]
Yes, I have just added a few more conditions

---
In reply to: [724948049](https://github.com/microsoft/onnxruntime/pull/5743#issuecomment-724948049) [](ancestors = 724948049)

File: csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs [hunk 2]
```diff
@@ -50,45 +50,65 @@ public void Dispose()
         #endregion
     }
 
+    /// <summary>
+    /// This class owns an OrtValue that
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs:54]: > an OrtValue that  [](start = 23, length = 18)

Complete the doc

File: csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs [hunk 3]
```diff
@@ -50,45 +50,65 @@ public void Dispose()
...
+        /// <returns>An instance of OrtValue that does not own underlying memory</returns>
         internal override OrtValue ToOrtValue(out MemoryHandle? pinnedMemoryHandle)
         {
             // PinnedMemoryHandle holds the default value as DisposableNamedOnnxValue
             // doesn't hold any managed buffer (that needs to be pinned)
             pinnedMemoryHandle = null;
-            // Assign the onnxValue by querying this instance's NativeOnnxTensorMemory instance
-            return new OrtValue(_nativeMemoryManager.Handle, false);
+            // Return non-owning instance of OrtValue
+            return _ortValueHolder.Value;
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs:95]: >_ortValueHolder [](start = 19, length = 15)

Throw when _ortValueHolder is null

**Review by RyanUnderhill — DISMISSED** [review]

**Review by hariharans29 — APPROVED** [review]

**Review by RyanUnderhill — APPROVED** [review]

**Pull request merged**